src/v/cluster/cluster_recovery_manager.cc

```diff
@@ -0,0 +1,201 @@
+  ss::sharded<controller_stm>& controller_stm,
+  ss::sharded<features::feature_table>& feature_table,
+  ss::sharded<cloud_storage::remote>& remote,
+  ss::sharded<cluster_recovery_table>& recovery_table,
+  ss::sharded<storage::api>& storage,
+  consensus_ptr raft0)
+  : _sharded_as(sharded_as)
+  , _controller_stm(controller_stm)
+  , _feature_table(feature_table)
+  , _remote(remote)
+  , _recovery_table(recovery_table)
+  , _storage(storage)
+  , _raft0(std::move(raft0)) {}
+
+ss::future<std::optional<model::term_id>>
+cluster_recovery_manager::sync_leader(ss::abort_source& as) {
+    if (!_raft0->is_leader()) {
+        co_return std::nullopt;
+    }
+    auto synced_term = _raft0->term();
+    // Make sure we catch up to the committed offset.
+    const auto committed_offset = _raft0->committed_offset();
+    auto res = co_await _raft0->linearizable_barrier();
+    // If we couldn't heartbeat to a majority, or the term has changed, exit
+    // out of the loop.
+    if (
+      !res.has_value() || !_raft0->is_leader()
+      || synced_term != _raft0->term()) {
+        co_return std::nullopt;
+    }
+    co_await _controller_stm.local().wait(
+      committed_offset, model::timeout_clock::time_point::max(), as);
+    if (!_raft0->is_leader() || synced_term != _raft0->term()) {
+        co_return std::nullopt;
+    }
+    co_return synced_term;
+}
+
+ss::future<result<cluster::errc, cloud_metadata::error_outcome>>
+cluster_recovery_manager::initialize_recovery(
+  cloud_storage_clients::bucket_name bucket) {
+    if (!_remote.local_is_initialized()) {
+        co_return cluster::errc::invalid_request;
+    }
+    if (_recovery_table.local().is_recovery_active()) {
+        co_return cluster::errc::cluster_already_exists;
+    }
+    if (
+      !_raft0->is_leader()
+      || !_storage.local().get_cluster_uuid().has_value()) {
+        co_return cluster::errc::not_leader_controller;
+    }
+    const auto& ignored_uuid = _storage.local().get_cluster_uuid().value();
+    // TODO: protect with semaphore with a term check.
+    auto fib = retry_chain_node{_sharded_as.local(), 30s, 1s};
+    auto cluster_manifest_res
+      = co_await cluster::cloud_metadata::download_highest_manifest_in_bucket(
+        _remote.local(), bucket, fib, ignored_uuid);
+    if (cluster_manifest_res.has_error()) {
+        vlog(
+          clusterlog.warn,
+          "Error finding cluster manifests in bucket {}: {}",
+          bucket,
+          cluster_manifest_res.error());
+        co_return cluster_manifest_res.error();
+    }
+    auto& manifest = cluster_manifest_res.value();
+    vlog(clusterlog.info, "Found cluster manifest for recovery: {}", manifest);
+
+    // Replicate an update to start recovery. Once applied, this will update
+    // the recovery table.
+    cluster_recovery_init_cmd_data data;
+    data.manifest = std::move(manifest);
+    data.bucket = std::move(bucket);
+    auto errc = co_await replicate_and_wait(
```

Comment:
if we want to make sure that there were no leadership changes in between a sync and this `replicatea_and_wait` call we can pass a term here, making this a conditional replicate.

Reply:
i think we should do it i.e. replicate with `synced_term`